replicator.go

```diff
@@ -0,0 +1,310 @@
+	highestRequested uint64
+
+	// highest we have received
+	highestObserved *signerRoundOrSeq
+
+	// receivedQuorumRounds maps either sequences or rounds to quorum rounds
+	receivedQuorumRounds map[uint64]QuorumRound
+
+	// request iterator
+	requestIterator int
+
+	timeoutHandler *TimeoutHandler
+}
+
+func newReplicator(logger Logger, sender sender, ourNodeID NodeID, maxRoundWindow uint64, start time.Time, lock *sync.Mutex) *replicator {
+	r := &replicator{
+		receivedQuorumRounds: make(map[uint64]QuorumRound),
+		sender:               sender,
+		myNodeID:             ourNodeID,
+		logger:               logger,
+		maxRoundWindow:       maxRoundWindow,
+		epochLock:            lock,
+	}
+
+	r.timeoutHandler = NewTimeoutHandler(logger, start, DefaultReplicationRequestTimeout, r.resendReplicationRequests)
+	return r
+}
+
+func (r *replicator) advanceTime(now time.Time) {
+	r.timeoutHandler.Tick(now)
+}
+
+func (r *replicator) resendReplicationRequests(missingIds []uint64) {
+	// we call this function in the timeout handler goroutine, so we need to
+	// ensure we don't have concurrent access to highestObserved
+	r.epochLock.Lock()
+	defer r.epochLock.Unlock()
+
+	nodes := r.highestObserved.signers
+	numNodes := len(nodes)
+	slices.Sort(missingIds)
+	segments := CompressSequences(missingIds)
+	for i, seqs := range segments {
+		index := (i + r.requestIterator) % numNodes
+		r.sendRequestToNode(seqs.Start, seqs.End, nodes[index])
+	}
+
+	r.requestIterator++
+}
+
+// isReplicationComplete returns true if we have finished the replication process.
+// The process is considered finished once highestObserved has caught up to the target
+// (either nextSeqToCommit or currentRound).
+func (r *replicator) isReplicationComplete(target uint64) bool {
+	if r.highestObserved != nil && r.highestObserved.value() >= target {
+		return false
+	}
+	return true
+}
+
+func (r *replicator) getHighestRound() uint64 {
+	if r.highestObserved != nil {
+		return r.highestObserved.round
+	}
+	return 0
+}
+
+// maybeSendMoreReplicationRequests checks if we need to send more replication requests given an observed round or sequence.
+// it limits the amount of outstanding requests to be at most [maxRoundWindow] ahead of [base] which is
+// either nextSeqToCommit or currentRound depending on if we are replicating sequences or rounds.
+func (r *replicator) maybeSendMoreReplicationRequests(observed *signerRoundOrSeq, base uint64) {
+	val := observed.value()
+
+	// we've observed something we've already requested
+	if r.highestRequested >= val && r.highestObserved != nil {
+		r.logger.Debug("Already requested observed value, skipping", zap.Uint64("value", val), zap.Bool("isRound", observed.isRound))
+		return
+	}
+
+	// if this is the highest observed sequence or round, update our state
+	if r.highestObserved == nil || val > r.highestObserved.value() {
+		r.highestObserved = observed
+	}
+
+	startSeq := math.Max(float64(base), float64(r.highestRequested))
```

Comment:
I don't understand - if `highestRequested` is a round, how can we treat it as a sequence? 

Can we perhaps maintain separate highestObservedX for rounds and sequences? I think it would make things more safer.